public/app/features/dashboard/components/GenAI/llms/openai.ts

```diff
@@ -0,0 +1,394 @@
+   *  - null: API response still in progress or incomplete
+   */
+  finish_reason: string;
+  /** The index of the completion in the list of choices. */
+  index: number;
+}
+
+/** The usage statistics for a request to OpenAPI. */
+export interface Usage {
+  /** The number of tokens in the prompt. */
+  prompt_tokens: number;
+  /** The number of tokens in the completion. */
+  completion_tokens: number;
+  /** The total number of tokens. */
+  total_tokens: number;
+}
+
+/** The error response from the Grafana LLM app when trying to call the chat completions API. */
+interface ChatCompletionsErrorResponse {
+  /** The error message. */
+  error: string;
+}
+
+/** A response from the OpenAI Chat Completions API. */
+export interface ChatCompletionsResponse<T = Choice> {
+  /** The ID of the request. */
+  id: string;
+  /** The type of object returned (e.g. 'chat.completion'). */
+  object: string;
+  /** The timestamp of the request, as a UNIX timestamp. */
+  created: number;
+  /** The name of the model used to generate the response. */
+  model: string;
+  /** A list of completion objects (only one, unless `n > 1` in the request). */
+  choices: T[];
+  /** The number of tokens used to generate the replies, counting prompt, completion, and total. */
+  usage: Usage;
+}
+
+/** A content message returned from the model. */
+export interface ContentMessage {
+  /** The content of the message. */
+  content: string;
+}
+
+/** A message returned from the model indicating that it is done. */
+export interface DoneMessage {
+  done: boolean;
+}
+
+/** A function call message returned from the model. */
+export interface FunctionCallMessage {
+  /** The name of the function to call. */
+  name: string;
+  /** JSON string for the arguments to the function call. */
+  arguments: string;
+}
+
+/**
+ * A delta returned from a stream of chat completion responses.
+ *
+ * In practice this will be either a content message or a function call;
+ * done messages are filtered out by the `streamChatCompletions` function.
+ */
+export type ChatCompletionsDelta = ContentMessage | FunctionCallMessage | DoneMessage;
+
+/** A chunk included in a chat completion response. */
+export interface ChatCompletionsChunk {
+  /** The delta since the previous chunk. */
+  delta: ChatCompletionsDelta;
+}
+
+/** Return true if the message is a 'content' message. */
+export function isContentMessage(message: unknown): message is ContentMessage {
+  return (message as ContentMessage).content != null;
+}
+
+/** Return true if the message is a 'done' message. */
+export function isDoneMessage(message: unknown): message is DoneMessage {
+  return (message as DoneMessage).done !== undefined;
+}
+
+/** Return true if the response is an error response. */
+export function isErrorResponse(response: unknown): response is ChatCompletionsErrorResponse {
+  return (response as ChatCompletionsErrorResponse).error !== undefined;
+}
```

Comment:
The linter doesn't like these functions and I cannot figure out any variation on them that it considers acceptable

Reply:
Similar PR here in experimental also not acceptable to the linter: https://github.com/grafana/grafana-experimental/pull/87

Reply:
Yep, I did some changes, but honestly, I don't know how to do better the type assertion here 😓